x-pack/plugins/fleet/server/services/fleet_server_migration.ts

```diff
@@ -63,6 +76,65 @@ function getInternalUserSOClient() {
   return appContextService.getInternalUserSOClient(fakeRequest);
 }
 
+async function migrateAgents() {
+  const esClient = appContextService.getInternalUserESClient();
+  const soClient = getInternalUserSOClient();
+  let hasMore = true;
+  while (hasMore) {
+    const res = await soClient.find({
+      type: AGENT_SAVED_OBJECT_TYPE,
+      page: 1,
+      perPage: 100,
+    });
+
+    if (res.total === 0) {
+      hasMore = false;
+    }
+    for (const so of res.saved_objects) {
+      try {
+        const {
+          attributes,
+        } = await appContextService
+          .getEncryptedSavedObjects()
+          .getDecryptedAsInternalUser<AgentSOAttributes>(AGENT_SAVED_OBJECT_TYPE, so.id);
+
+        const body: FleetServerAgent = {
+          type: attributes.type,
+          active: attributes.active,
+          enrolled_at: attributes.enrolled_at,
+          unenrolled_at: attributes.unenrolled_at,
+          unenrollment_started_at: attributes.unenrollment_started_at,
+          upgraded_at: attributes.upgraded_at,
+          upgrade_started_at: attributes.upgrade_started_at,
+          access_api_key_id: attributes.access_api_key_id,
+          user_provided_metadata: attributes.user_provided_metadata,
+          local_metadata: attributes.local_metadata,
+          policy_id: attributes.policy_id,
+          policy_revision_idx: attributes.policy_revision || undefined,
+          last_checkin: attributes.last_checkin,
+          last_checkin_status: attributes.last_checkin_status,
+          default_api_key_id: attributes.default_api_key_id,
+          default_api_key: attributes.default_api_key,
+          packages: attributes.packages,
+        };
+        await esClient.create({
+          index: AGENTS_INDEX,
+          body,
+          id: so.id,
+          refresh: 'wait_for',
+        });
+
+        await soClient.delete(AGENT_SAVED_OBJECT_TYPE, so.id);
```

Comment:
I guess bulk ES create is not possible since we need to decrypt the agent SO individually. but what do you think of doing a single bulk SO delete outside of this loop so that's more performant?

Reply:
unfortunately there is no `bulkDelete` on the saved object client

Reply:
For the performance aspect: I think we have a few things to mitigate that already in place with a pending state in the setup.

Also I am working on a PR that move the migration again from the setup call to the plugin start, where we will create the system indices, then start the migration but not block the plugin start.